### PR TITLE
Search Milestone 2: Add `verified` search filter

### PR DESCRIPTION
**Pull request by oisincoveney**
_**Note: this PR is blocked until the `Switch` Mantine component is available in the very near future.** Otherwise, the code is ready to be reviewed so have at it_ 😄

`verified` filter for Search Milestone 2: #27982 

### Description

Adds `verified` filter toggle to the search sidebar. The code only appears when using enterprise with the `content_verification` token enabled.

### How to verify

Start an enterprise instance with `yarn dev-ee` - ensure that you have an enterprise token (instructions are [here](https://www.notion.so/metabase/Cypress-How-To-4d1c70bb0ce94a60b3688fd7acba8471?pvs=4))

Go to the search sidebar on an enterprise instance, and filter by verified items only. Only verified assets should be available in the search after the filter.

After filtering, try to refresh the page or paste the URL into a new window to test that hydrating the search filters still works when entering the search directly from a URL. The results should remain the same and should behave in the same way.

### Tests

- [X] Tests have been added for all components

### CODE REVIEW

File: frontend/src/metabase/search/components/SearchSidebar/DropdownSidebarFilter/DropdownSidebarFilter.tsx [hunk 1]
```diff
@@ -4,7 +4,7 @@ import { useLayoutEffect, useRef, useState } from "react";
 import { t } from "ttag";
```
Comment (oisincoveney) [inline frontend/src/metabase/search/components/SearchSidebar/DropdownSidebarFilter/DropdownSidebarFilter.tsx:4]: This file and the next are just changing some type names and things so I can have a toggle filter (which we'll also need for the SQL searching)

File: frontend/src/metabase/search/components/SearchSidebar/SearchSidebar.tsx [hunk 2]
```diff
@@ -22,11 +21,37 @@ export const SearchSidebar = ({
...
+  const filterMap: Record<FilterTypeKeys, SearchFilterComponent | null> =
+    useMemo(
+      () => ({
+        [SearchFilterKeys.Type]: TypeFilter,
+        [SearchFilterKeys.Verified]: PLUGIN_CONTENT_VERIFICATION.VerifiedFilter,
+      }),
+      [],
+    );
+
+  const isValidFilterValue = useCallback(
```
Comment (oisincoveney) [inline frontend/src/metabase/search/components/SearchSidebar/SearchSidebar.tsx:33]: Might move this to a `validate` function on a per-filter basis in the next iteration. Let me know what you think, maybe it should be done now.

File: e2e/test/scenarios/onboarding/search/search.cy.spec.js [hunk 3]
```diff
@@ -1,6 +1,7 @@
 import _ from "underscore";
```
Comment (oisincoveney) [inline e2e/test/scenarios/onboarding/search/search.cy.spec.js:1]: There was an absolute merge mess here due to the addition of `describeEE` and a reorganization to make sure that each filter had its own set of tests. Feel free to just look at the raw file or the unified diff for this one

File: frontend/src/metabase/search/components/SearchSidebar/DropdownSidebarFilter/DropdownSidebarFilter.tsx [hunk 4]
```diff
@@ -2,8 +2,10 @@ import { isEmpty } from "underscore";
 import type { MouseEvent } from "react";
```
Comment (oisincoveney) [inline frontend/src/metabase/search/components/SearchSidebar/DropdownSidebarFilter/DropdownSidebarFilter.tsx:2]: I rename the `SidebarFilter` type to `DropdownSidebarFilter` to reflect the addition of a `ToggleSidebarFilter` type. Most of these changes are just renaming

**Review by ranquild — APPROVED** [review]
I don't like that search filters are implemented in this "generic" way. I'd prefer it to be very simple and plain, maybe with some code repetition. This looks like a small framework.  

Still, this PR on its own is all right I think

File: enterprise/frontend/src/metabase-enterprise/content_verification/index.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+import { PLUGIN_CONTENT_VERIFICATION } from "metabase/plugins";
+import { hasPremiumFeature } from "metabase-enterprise/settings";
+import { VerifiedFilter } from "metabase-enterprise/content_verification/VerifiedFilter/VerifiedFilter";
```
Comment (WiNloSt) [inline enterprise/frontend/src/metabase-enterprise/content_verification/index.ts:3]: nit: what do you think about adding `.../VerifiedFilter/index.ts` so we don't need to see `VerifiedFilter` being repeated when importing the component?

File: enterprise/frontend/src/metabase-enterprise/content_verification/VerifiedFilter/VerifiedFilter.tsx [hunk 6]
```diff
@@ -0,0 +1,9 @@
+import { t } from "ttag";
+import type { SearchFilterComponent } from "metabase/search/types";
+
+export const VerifiedFilter: SearchFilterComponent<"verified"> = {
+  title: t`Verified items only`,
+  type: "toggle",
+  fromUrl: value => value === "true",
+  toUrl: (value: boolean) => (value ? "true" : undefined),
```
Comment (WiNloSt) [inline enterprise/frontend/src/metabase-enterprise/content_verification/VerifiedFilter/VerifiedFilter.tsx:8]: Can you explain a little about these filter APIs? i.e. how `fromUrl` and `toUrl` work. I figure it might have something to do with hydrating the value from the URL and slapping the value back to the URL when the filter changes.

File: frontend/src/metabase/search/components/SearchSidebar/SearchSidebar.tsx [hunk 7]
```diff
@@ -1,59 +1,104 @@
...
   onChange: (value: URLSearchFilterQueryParams) => void;
 };
 
-export const filterMap: Record<FilterTypeKeys, SearchSidebarFilterComponent> = {
-  [SearchFilterKeys.Type]: TypeFilter,
-  [SearchFilterKeys.CreatedBy]: CreatedByFilter,
-};
-
 export const SearchSidebar = ({ value, onChange }: SearchSidebarProps) => {
+  const filterMap: Record<FilterTypeKeys, SearchFilterComponent | null> =
```
Comment (WiNloSt) [inline frontend/src/metabase/search/components/SearchSidebar/SearchSidebar.tsx:23]: Could/should we move this map outside of the component so we don't need to use `useMemo` since it doesn't rely on any props?

File: frontend/src/metabase/search/components/SearchSidebar/SearchSidebar.tsx [hunk 8]
```diff
@@ -1,59 +1,104 @@
...
+        [SearchFilterKeys.CreatedBy]: CreatedByFilter,
+        [SearchFilterKeys.Verified]: PLUGIN_CONTENT_VERIFICATION.VerifiedFilter,
+      }),
+      [],
+    );
+
+  const isValidFilterValue = useCallback(
+    (
+      key: FilterTypeKeys,
+      val: SearchFilterPropTypes[FilterTypeKeys],
```
Comment (WiNloSt) [inline frontend/src/metabase/search/components/SearchSidebar/SearchSidebar.tsx:36]: can you use the unabbreviated version `value` instead? We try not to abbreviate names as much as possible to make everything as clear as possible.

edited: I noticed this was kind of needed because we also have another variable called `value`, but to be fare, both names sound pretty similar but have totally different shapes. I think we should differentiate them more e.g.
- `value` -> `filters` since this is a map that contains all the filters.
- `val` -> `value` since this is filter values e.g. string or array

File: frontend/src/metabase/search/components/SearchSidebar/SearchSidebar.tsx [hunk 9]
```diff
@@ -1,59 +1,104 @@
...
     key: FilterTypeKeys,
     val: SearchFilterPropTypes[FilterTypeKeys],
   ) => {
-    if (!val || (Array.isArray(val) && val.length === 0)) {
+    if (!isValidFilterValue(key, val)) {
       onChange(_.omit(value, key));
     } else {
-      const { toUrl } = filterMap[key];
+      const filterMapElement = filterMap[key];
+      const toUrl = filterMapElement?.toUrl;
```
Comment (WiNloSt) [inline frontend/src/metabase/search/components/SearchSidebar/SearchSidebar.tsx:58]: Is this still needed even though we filter out invalid filter values on line 54? Feels to me `filterMapElement` at this point shouldn't be nullable.

Or is this change needed just to satisfy TypeScript? If that's the case, I think we should fix `isValidFilterValue` to let TypeScript know it's filtering out invalid values instead.

File: frontend/src/metabase/search/components/SearchSidebar/SearchSidebar.tsx [hunk 10]
```diff
@@ -1,59 +1,104 @@
...
-        filter={Filter}
-        data-testid={`${key}-search-filter`}
-        value={filterValue}
-        onChange={value => onOutputChange(key, value)}
-      />
-    );
+
+    if (!Filter) {
+      return null;
+    }
```
Comment (WiNloSt) [inline frontend/src/metabase/search/components/SearchSidebar/SearchSidebar.tsx:71]: one way to simplify this is to have a default value of the component from enterprise folder being a component that renders nothing instead of null e.g.
```
() => null
```

This will ensure we don't need a special code to deal with it and it should simplify its relevant parts quite a bit too.

**Review by WiNloSt — COMMENTED** [review]
This is 1/3 of the review. I'll submit more later. Gotta rest 😵‍💫

**Review by WiNloSt — APPROVED** [review]
I have 2 more comments regarding UI and UX.
1. The search filter isn't aligned with the search results, unlike the mockup on Figma.
    Screenshot 2023-09-27 at 4 16 02 PM
2. After modifying any filters e.g. check the verified filter, if I want to search a different name, I would go to the search bar type something else and search again, this results in all filters being reverted to default values.

File: enterprise/frontend/src/metabase-enterprise/content_verification/VerifiedFilter/VerifiedFilter.tsx (hunk 6 above)
Comment (oisincoveney) [reply enterprise/frontend/src/metabase-enterprise/content_verification/VerifiedFilter/VerifiedFilter.tsx:8]: Yeah, I left the descriptions of what they do in the type definition so I don't pollute all of the comments 😄

File: frontend/src/metabase/search/components/filters/CreatedByFilter/CreatedByContent.tsx [hunk 11]
```diff
@@ -9,40 +9,45 @@ import {
...
-import type { SearchSidebarFilterComponent } from "metabase/search/types";
+import type { SearchFilterDropdown } from "metabase/search/types";
 import { UserListElement } from "metabase/search/components/filters/CreatedByFilter/UserListElement";
-import { SearchFilterPopoverWrapper } from "metabase/search/components/SidebarFilter/SearchFilterPopoverWrapper";
+import { SearchFilterPopoverWrapper } from "metabase/search/components/SearchSidebar/DropdownSidebarFilter/SearchFilterPopoverWrapper";
 
-export const CreatedByContent: SearchSidebarFilterComponent<"created_by">["ContentComponent"] =
-  ({ value, onChange, onApply }) => {
+export const CreatedByContent: SearchFilterDropdown<"created_by">["ContentComponent"] =
+  ({ value, onChange }) => {
```
[the remark this answers is not in the document]
Comment (oisincoveney) [reply frontend/src/metabase/search/components/filters/CreatedByFilter/CreatedByContent.tsx:17]: I was struggling with the semantics on this one for a bit - with this previous `onApply` logic, I was trying to differentiate between "changing the filters" and "actually applying the filters" and where I should put that logic. I was thinking that I would have each content component tell the parent whether the filter should be applied or not.

I came to the conclusion that it's silly to have these two states, and that the content components should really just send the final result back to the parent component while keeping the intermediate (unapplied) states within the children components.

Let me know if that makes sense 😄

File: frontend/src/metabase/search/components/SearchSidebar/ToggleSidebarFilter/ToggleSidebarFilter.tsx [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+      </Text>
+      <Switch
+        wrapperProps={{
+          "data-testid": "toggle-filter-switch",
+        }}
+        size="sm"
+        data-is-checked={booleanValue}
+        checked={booleanValue}
+        onChange={event =>
+          onChange(event.currentTarget.checked ? true : undefined)
```
[the remark this answers is not in the document]
Comment (oisincoveney) [reply frontend/src/metabase/search/components/SearchSidebar/ToggleSidebarFilter/ToggleSidebarFilter.tsx:32]: Yeah, if this component is going to be generic then it should return a generic response, and the Verified filter should handle the `true` vs `undefined` case.

File: frontend/src/metabase/search/components/SearchSidebar/ToggleSidebarFilter/ToggleSidebarFilter.unit.spec.tsx [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+
+const MockToggleSidebarFilter = ({
+  filter,
+  value,
+  onChange,
+}: ToggleSidebarFilterProps) => {
+  const [toggleValue, setToggleValue] = useState(value);
+  const onFilterChange = (elem: ToggleSidebarFilterProps["value"]) => {
+    setToggleValue(elem);
+    onChange(elem);
```
[the remark this answers is not in the document]
Comment (oisincoveney) [reply frontend/src/metabase/search/components/SearchSidebar/ToggleSidebarFilter/ToggleSidebarFilter.unit.spec.tsx:25]: Changed to `toggleValue` for extreme explicitness 😄

**Comment by oisincoveney** [comment]
> I have 2 more comments regarding UI and UX.
> 
> 1. The search filter isn't aligned with the search results, unlike the mockup on Figma.

I've sorted that out, should look good now.

> 2. After modifying any filters e.g. check the verified filter, if I want to search a different name, I would go to the search bar type something else and search again, this results in all filters being reverted to default values.

I think this is something that we can bring up in a bigger discussion on Slack, since this is more of a UX flow thing that I think the designers and product people have more interesting perspectives on 😄

File: frontend/src/metabase/search/components/SearchSidebar/ToggleSidebarFilter/ToggleSidebarFilter.tsx [hunk 14]
```diff
@@ -1,21 +1,50 @@
...
+ *
+ *
+ *
+ *
+ *
+ *
+ *
+ *
+ *
+ * TODO: Change mantine switch import before merging!!
```
Comment (WiNloSt) [inline frontend/src/metabase/search/components/SearchSidebar/ToggleSidebarFilter/ToggleSidebarFilter.tsx:13]: 😂 This is one hell of a way to make sure we don't forget to fix this.

File: frontend/src/metabase/search/components/filters/CreatedByFilter/UserListElement/UserListElement.unit.spec.tsx [hunk 15]
```diff
@@ -1,7 +1,7 @@
 import userEvent from "@testing-library/user-event";
 import { renderWithProviders, screen } from "__support__/ui";
 import { createMockUserListResult } from "metabase-types/api/mocks";
-import { UserListElement } from "metabase/search/components/filters/CreatedByFilter/UserListElement";
+import { UserListElement } from "metabase/search/components/filters/CreatedByFilter/UserListElement/UserListElement";
```
Comment (WiNloSt) [inline frontend/src/metabase/search/components/filters/CreatedByFilter/UserListElement/UserListElement.unit.spec.tsx:4]: Ahhh. I think this might be the remnant from refactoring, you should be able to leave this as is (revert the change back)

File: frontend/src/metabase/search/components/SearchSidebar/ToggleSidebarFilter/ToggleSidebarFilter.tsx (hunk 14 above)
Comment (oisincoveney) [reply frontend/src/metabase/search/components/SearchSidebar/ToggleSidebarFilter/ToggleSidebarFilter.tsx:13]: If it's loud enough I won't forget 😄

File: frontend/src/metabase/search/components/filters/CreatedByFilter/UserListElement/UserListElement.unit.spec.tsx (hunk 15 above)
Comment (oisincoveney) [reply frontend/src/metabase/search/components/filters/CreatedByFilter/UserListElement/UserListElement.unit.spec.tsx:4]: Smh. The auto-import from Webstorm strikes again
